WcaOnRails/app/controllers/competitions_controller.rb

```diff
@@ -278,6 +278,20 @@ def update_events
     end
   end
 
+  def update_wcif_events
+    @competition = competition_from_params
+    wcif_events = params["_json"].map { |wcif_event| wcif_event.permit!.to_h } #<<< TODO: validate that the incoming stuff is actually wcif >>>
```

Comment:
I had a few thoughts bouncing around in my head, all of which you touched on in your comment:

1. People are more clever than me, and while I cannot think of any ways to craft a malicious WCIF that does anything bad, I also did not want to think throught it. In a past implementation of this, the code would actually call a `send(...)` based on the WCIF given by the user, which felt eerily close to `eval`-ing user submitted code. We have since redefined the `AdvancementCondition` so that is no longer the case.
2. With the introduction of the nuke and replace implementation, I found that any error in the WCIF document would cause all the rounds for an event to get nuked. This seemed like bad behavior to me: the update should either happen in its entirety, or nothing should happen. At first this felt like a problem that would be fixed by validating the WCIF document up front, but then I realized there's *always* a chance that something could go wrong later on. The right fix for this was to add a [test](https://github.com/jfly/worldcubeassociation.org/blob/8f721ea6fcbe42414011428d9e9e19bec647abab/WcaOnRails/spec/requests/competitions_spec.rb#L63-L93) and [wrap the WCIF update in a database transaction](https://github.com/jfly/worldcubeassociation.org/blob/8f721ea6fcbe42414011428d9e9e19bec647abab/WcaOnRails/app/models/competition.rb#L828).
3. When developing the UI that talks to this api, I was running into 2) and getting some weird errors. I thought an upfront validation of the *shape* (or structure, as you said) of the document could give much nicer error messages.

I'm going to experiment with the JSON schema thing to see if it is a pain or not.